src/lib.rs

```diff
@@ -1,374 +1,47 @@
-
+mod destination;
 mod fs_watcher;
+mod identity;
 mod spec;
 
-pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
-
-#[derive(Clone, Debug)]
-pub struct DstService {
-    inner: Arc<Inner>,
-}
-
-#[derive(Debug)]
-pub struct DstSender {
-    endpoints: HashMap<Dst, watch::Sender<Endpoints>>,
-    overrides: HashMap<Dst, watch::Sender<Overrides>>,
-    inner: Weak<Inner>,
-}
-
-impl DstSender {
-    #[tracing::instrument(skip(self), name = "DstSender::send_endpoints", level = "info")]
-    pub async fn send_endpoints(&mut self, dst: Dst, endpoints: Endpoints) -> Result<(), Error> {
-        if let Some(sender) = self.endpoints.get(&dst) {
-            tracing::info!("Dst present");
-            sender.broadcast(endpoints)?;
-        } else {
-            tracing::info!("Dst non present");
-            if let Some(inner) = self.inner.upgrade() {
-                let (tx, rx) = watch::channel(endpoints);
-                self.endpoints.insert(dst.clone(), tx);
-                inner.endpoints.write().await.insert(dst, rx);
-            }
-        }
-        Ok(())
-    }
-
-    #[tracing::instrument(skip(self), name = "DstSender::delete_dst", level = "info")]
-    pub async fn delete_dst(&mut self, dst: Dst) {
-        if let Some(sender) = self.endpoints.remove(&dst) {
-            tracing::info!("dropping sender");
-            drop(sender);
-            if let Some(inner) = self.inner.upgrade() {
-                inner.endpoints.write().await.remove(&dst);
-            }
-        } else {
-            tracing::info!("Dst not found");
-        }
-    }
-}
-
-#[derive(Debug)]
-pub struct Inner {
-    endpoints: RwLock<HashMap<Dst, watch::Receiver<Endpoints>>>,
-    overrides: RwLock<HashMap<Dst, watch::Receiver<Overrides>>>,
-}
-
-#[derive(Debug, PartialEq, Eq, Default, Clone)]
-pub struct Endpoints(HashMap<SocketAddr, EndpointMeta>);
+pub use self::destination::{Dst, DstSender, DstService, EndpointMeta, Endpoints, Overrides};
+pub use self::fs_watcher::FsWatcher;
+pub use self::identity::IdentityService;
+pub use self::spec::{EndpointsSpec, OverridesSpec};
 
-#[derive(Debug, Hash, PartialEq, Eq, Clone, Serialize, Deserialize)]
-pub struct EndpointMeta {
-    address: SocketAddr,
-    h2: bool,
-    weight: u32,
-    #[serde(default)]
-    metric_labels: BTreeMap<String, String>,
-    tls_identity: Option<String>,
-    authority_override: Option<String>,
-}
+use linkerd2_proxy_api::{
+    destination::destination_server::DestinationServer, identity::identity_server::IdentityServer,
+};
+use std::net::SocketAddr;
+use tracing_futures::Instrument;
 
-#[derive(Debug, PartialEq, Eq, Default, Clone)]
-struct Overrides(HashMap<Dst, u32>);
+pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
 
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub struct Dst {
-    name: String,
-    port: u16,
+pub struct Controller {
+    dst_svc: DstService,
+    identity_svc: IdentityService,
```

Comment:
I feel like this project could now be better named `linkerd2-mock-controller`, but I didn't make that change as part of this. This is responsible for actually serving the destination and identity services!